### PR TITLE
add connector sakuranovel

### LINKED ISSUE #4185: [Site Request] Sakuranovel
**Name of the website**
Sakuranovel

**Website urls (examples below)**
- Site: https://sakuranovel.id/
- Manga List: https://sakuranovel.id/daftar-novel/
- Manga example: https://sakuranovel.id/series/overlimit-skill-holder/
- Chapter online viewer example: https://sakuranovel.id/fake-saint-of-the-year-side-story-2b-bahasa-indonesia/

**Languages**
Indonesian

**Website relationship*
If applicable describe the relation with any other website (eg : "alternative domain, copy of ...").

**Additional details**
Add any other context details that you may have found (like template to reuse)

### PR DESCRIPTION
**Pull request by akn19**
fixes : https://github.com/manga-download/hakuneko/issues/4185, https://github.com/manga-download/hakuneko/issues/5734

### CODE REVIEW

**Comment by MikeZeDev** [comment]
How did i miss this dupe?

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        this.tags = ['novel', 'indonesian'];
+        this.url = 'https://sakuranovel.id';
+
+        this.querMangaTitleFromURI = 'div.series-title h2';
+        this.queryMangas = 'div.novellist-blc ul li a.series';
+        this.queryChapters = 'ul.series-chapterlists li a';
+        this.queryChaptersTitle = 'span:not(.date)';
+        this.path = '/daftar-novel/';
+
+        this.novelContentQuery = 'div.asdasd';
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs:18]: I have the feeling that this selector is not good

Edit : well it exists but there is aproblem

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        this.novelFormat = 'image/png';
+        this.novelWidth = '56em'; // parseInt(1200 / window.devicePixelRatio) + 'px';
+        this.novelPadding = '1.5em';
+    }
+
+    async _getChapters(manga) {
+        let request = new Request(new URL(manga.id, this.url), this.requestOptions);
+        let data = await this.fetchDOM(request, this.queryChapters);
+        return data.map(element => {
+            this.adLinkDecrypt(element);
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs]: useless

File: src/web/mjs/connectors/SakuraNovel.mjs (hunk 1 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/SakuraNovel.mjs:18]: image

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            const title = this.queryChaptersTitle ? [...element.querySelectorAll(this.queryChaptersTitle)].map(ele => ele.textContent).join(' ') : element.text;
+            return {
+                id: this.getRootRelativeOrAbsoluteLink(element, request.url),
+                title: title.replace(manga.title + ' – ', '').replace(/Bahasa Indonesia$/i, '').trim()
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs]: I ask myself why do they need to put the language everywhere.. More garbage, more and mooore garbage..
Anyway,  chapters still got manga title for some reasons according to the screenshot i took.

Did you test? Its enough to use super._getchapters() and then remove bahasa, EXACTLY like you did on monzee for mangas. No need for that complicated stuff.

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            };
+        });
+    }
+
+    async _getPages(chapter) {
+        let uri = new URL(chapter.id, this.url);
+        uri.searchParams.set('style', 'list');
+        let request = new Request(uri, this.requestOptions);
+        let data = await this.fetchDOM(request, this.novelContentQuery);
+        return data.length > 0 ? this._getPagesNovel(request) : super._getPages(chapter);
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs]: If they have only novels its useless to check you can just call _getPagesNovel.

**Review by MikeZeDev — CHANGES_REQUESTED** [review]

File: src/web/mjs/connectors/SakuraNovel.mjs (hunk 1 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/SakuraNovel.mjs:18]: Please test your code before submitting, its not working.

**Comment by akn19** [comment]
the code I used previously came from the KolNovel connector.

Excessive use of the word "Bahasa Indonesia" is most likely aimed at SEO, because of the low literacy of our society, they will be more likely to visit websites with titles containing the word "Bahasa Indonesia" in search engines.

**Comment by akn19** [comment]
gambar

File: src/web/mjs/connectors/SakuraNovel.mjs (hunk 1 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/SakuraNovel.mjs:18]: Well its better the thing is i wonder why i works on some novels and not on others, like i see in the screen.

🤔 

It doesnt render anything on "Academys black haired foreigner" or "I became a g"nius commander..." chapter
i guess the script should be tweaked a bit?

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return this._getPagesNovel(request);
+    }
+
+    async _getPagesNovel(request) {
+        let darkmode = Engine.Settings.NovelColorProfile();
+        let script = `
+            new Promise((resolve, reject) => {
+                document.body.className = document.body.className.replace('darkmode', 'lightmode')
+                document.body.style.width = '${this.novelWidth}';
+                let container = document.querySelector('div.asdasd');
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs]: you made a constant for this, you may want to use it there :)

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+        this.novelContentQuery = 'div.asdasd';
+        this.novelObstaclesQuery = 'div[class]';
+        this.novelFormat = 'image/png';
+        this.novelWidth = '56em'; // parseInt(1200 / window.devicePixelRatio) + 'px';
+        this.novelPadding = '1.5em';
+    }
+
+    async _getChapters(manga) {
+        const mangas = await super._getChapters(manga);
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs]: Suggestion to use relevant variables names, we are not gathering mangas here :)

**Review by MikeZeDev — CHANGES_REQUESTED** [review]
Minor changes required, and we must know why it doesnt work everywhere likz in the screen i posted

File: src/web/mjs/connectors/SakuraNovel.mjs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        let uri = new URL(chapter.id, this.url);
+        let request = new Request(uri, this.requestOptions);
+        return this._getPagesNovel(request);
+    }
+
+    async _getPagesNovel(request) {
+        let darkmode = Engine.Settings.NovelColorProfile();
+        let script = `
+            new Promise((resolve, reject) => {
+                document.body.className = document.body.className.replace('darkmode', 'lightmode')
```
Comment (MikeZeDev) [inline src/web/mjs/connectors/SakuraNovel.mjs]: lightmode to replace darkmode? why this line? its that from elsewhere? remove it perhaps.

File: src/web/mjs/connectors/SakuraNovel.mjs (hunk 1 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/SakuraNovel.mjs:18]: I get it. Script remove nodes matched with novelObstaclesQuery. We have legitimate text nodes <div class ="text-align"> that are removed. Suggestion to remove google ads instead : 

        this.novelObstaclesQuery = 'div[data-index]';

**Comment by akn19** [comment]
content errors fixed.

I have tried it on several titles and so far no issues have arisen.

File: src/web/mjs/connectors/SakuraNovel.mjs (hunk 7 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/SakuraNovel.mjs]: this still need to be addressed :)

**Comment by MikeZeDev** [comment]
For the integration to work you'll also have to update your branch from our master. Github updated lint so we had to lock the version.

File: src/web/mjs/connectors/SakuraNovel.mjs (hunk 7 above)
Comment (MikeZeDev) [reply src/web/mjs/connectors/SakuraNovel.mjs]: @akn19 please remove the useless darkode/lightmode line and we are settled <3

**Review by MikeZeDev — APPROVED** [review]

**Pull request merged**